providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+# define ARGON2_MIN_AD_LENGTH UINT32_C(0)
+# define ARGON2_MAX_AD_LENGTH UINT32_C(0xFFFFFFFF)
+# define ARGON2_MIN_SALT_LENGTH UINT32_C(8)
+# define ARGON2_MAX_SALT_LENGTH UINT32_C(0xFFFFFFFF)
+# define ARGON2_MIN_SECRET UINT32_C(0)
+# define ARGON2_MAX_SECRET UINT32_C(0xFFFFFFFF)
+# define ARGON2_DEFAULT_FLAGS UINT32_C(0)
+# define ARGON2_FLAG_CLEAR_PASSWORD (UINT32_C(1) << 0)
+# define ARGON2_FLAG_CLEAR_SECRET (UINT32_C(1) << 1)
+# define ARGON2_BLOCK_SIZE 1024
+# define ARGON2_QWORDS_IN_BLOCK ARGON2_BLOCK_SIZE / 8
+# define ARGON2_OWORDS_IN_BLOCK ARGON2_BLOCK_SIZE / 16
+# define ARGON2_HWORDS_IN_BLOCK ARGON2_BLOCK_SIZE / 32
+# define ARGON2_512BIT_WORDS_IN_BLOCK ARGON2_BLOCK_SIZE / 64
+# define ARGON2_ADDRESSES_IN_BLOCK 128
+# define ARGON2_PREHASH_DIGEST_LENGTH 64
+# define ARGON2_PREHASH_SEED_LENGTH 72
+
+# undef G
+# define G(a, b, c, d)                                                         \
+    do {                                                                       \
+        a = a + b + 2 * mul_lower(a, b);                                       \
+        d = rotr64(d ^ a, 32);                                                 \
+        c = c + d + 2 * mul_lower(c, d);                                       \
+        b = rotr64(b ^ c, 24);                                                 \
+        a = a + b + 2 * mul_lower(a, b);                                       \
+        d = rotr64(d ^ a, 16);                                                 \
+        c = c + d + 2 * mul_lower(c, d);                                       \
+        b = rotr64(b ^ c, 63);                                                 \
+    } while ((void)0, 0)
+
+# undef BLAKE2_ROUND
+# define BLAKE2_ROUND(v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11,        \
+                       v12, v13, v14, v15)                                     \
+    do {                                                                       \
+        G(v0, v4, v8, v12);                                                    \
+        G(v1, v5, v9, v13);                                                    \
+        G(v2, v6, v10, v14);                                                   \
+        G(v3, v7, v11, v15);                                                   \
+        G(v0, v5, v10, v15);                                                   \
+        G(v1, v6, v11, v12);                                                   \
+        G(v2, v7, v8, v13);                                                    \
+        G(v3, v4, v9, v14);                                                    \
+    } while ((void)0, 0)
+
+# undef BLAKE2_ROUND_COLUMN
+# define BLAKE2_ROUND_COLUMN(x, i)                                             \
+    do {                                                                       \
+        BLAKE2_ROUND(                                                          \
+            x[16 * i],      x[16 * i + 1],  x[16 * i + 2],  x[16 * i + 3],     \
```

Comment:
I assume this gets optimized but you could maybe cache the 16*i bit (or put it in a pointer)

Reply:
Opted for the latter approach. Thanks!